src/recovery/mod.rs

```diff
@@ -236,13 +249,62 @@ impl Recovery {
             self.hystart.start_round(pkt_num);
         }
 
+        // Pacing: Set the pacing rate if BBR is not used
+        if !(self.cc_ops.has_custom_pacing)() {
+            if let Some(srtt) = self.smoothed_rtt {
+                let rate = (self.congestion_window as u64 * 1000000) /
+                    srtt.as_micros() as u64;
+                self.set_pacing_rate(rate);
+            }
+        }
+
+        self.schedule_next_packet(epoch, now, sent_bytes);
+
+        self.bytes_sent += sent_bytes;
         trace!("{} {:?}", trace_id, self);
     }
 
     fn on_packet_sent_cc(&mut self, sent_bytes: usize, now: Instant) {
         (self.cc_ops.on_packet_sent)(self, sent_bytes, now);
     }
 
+    pub fn set_pacing_rate(&mut self, rate: u64) {
+        if rate > 0 {
+            self.pacing_rate = rate;
+        }
+    }
+
+    pub fn get_packet_send_time(&self) -> Option<Instant> {
+        self.last_packet_scheduled_time
+    }
+
+    fn schedule_next_packet(
+        &mut self, epoch: packet::Epoch, now: Instant, packet_size: usize,
+    ) {
+        // Pacing is not done for following cases,
+        // 1. Packet epoch is not EPOCH_APPLICATION.
+        // 2. If packet has only ACK frames.
+        // 3. Start of the connection.
+        if epoch != packet::EPOCH_APPLICATION ||
+            packet_size == 0 ||
+            self.bytes_sent <= self.congestion_window ||
+            self.pacing_rate == 0
+        {
+            self.last_packet_scheduled_time = Some(now);
+            return;
+        }
+
+        let interval: u64 = (packet_size as u64 * 1000000) / self.pacing_rate;
+        let next_send_interval = Duration::from_micros(interval);
+
+        let next_schedule_time = match self.last_packet_scheduled_time {
+            Some(last_scheduled_time) => last_scheduled_time + next_send_interval,
+            None => now,
+        };
```

Comment:
Since you only use the interval in the `Some` case you could probably rewrite this as 

```
 self.last_packet_scheduled_time = match self.last_packet_scheduled_time {
    Some(last_scheduled_time) => {
        let interval: u64 = (packet_size as u64 * 1000000) / self.pacing_rate;
        let interval = Duration::from_micros(interval);
        let next_schedule_time = last_scheduled_time + interval;
        Some(cmp::max(now, next_schedule_time))
    }.

    None => Some(now),
 };
```